flink-end-to-end-tests/flink-end-to-end-tests-sql/src/test/java/org/apache/flink/table/sql/codegen/AddRemoteJarITCase.java

```diff
@@ -0,0 +1,280 @@
+    private org.apache.hadoop.fs.FileSystem hdfs;
+
+    public AddRemoteJarITCase(String executionMode) {
+        this.executionMode = executionMode;
+    }
+
+    @BeforeClass
+    public static void verifyOS() {
+        Assume.assumeTrue(
+                "HDFS cluster cannot be started on Windows without extensions.",
+                !OperatingSystem.isWindows());
+    }
+
+    @Before
+    public void createHDFS() {
+        try {
+            Configuration hdConf = new Configuration();
+
+            File baseDir = new File("./target/hdfs/hdfsTest").getAbsoluteFile();
+            FileUtil.fullyDelete(baseDir);
+            hdConf.set(MiniDFSCluster.HDFS_MINIDFS_BASEDIR, baseDir.getAbsolutePath());
+            MiniDFSCluster.Builder builder = new MiniDFSCluster.Builder(hdConf);
+            hdfsCluster = builder.build();
+
+            hdPath = new org.apache.hadoop.fs.Path("/test.jar");
+            hdfs = hdPath.getFileSystem(hdConf);
+
+            File localUdfJar =
+                    UserClassLoaderJarTestUtils.createJarFile(
+                            tempFolder.newFolder(String.format("test-jar-%s", UUID.randomUUID())),
+                            "test-classloader-udf.jar",
+                            GENERATED_LOWER_UDF_CLASS,
+                            String.format(GENERATED_LOWER_UDF_CODE, GENERATED_LOWER_UDF_CLASS));
+            hdfs.copyFromLocalFile(new org.apache.hadoop.fs.Path(localUdfJar.toURI()), hdPath);
+        } catch (Throwable e) {
+            e.printStackTrace();
+            Assert.fail("Test failed " + e.getMessage());
+        }
+
+        Path tmpPath = tempFolder.getRoot().toPath();
+        LOG.info("The current temporary path: {}", tmpPath);
+        this.result = tmpPath.resolve("result");
+    }
+
+    @After
+    public void destroyHDFS() {
+        try {
+            hdfs.delete(hdPath, false);
+            hdfsCluster.shutdown();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Test
+    public void testAddRemoteJar() throws Exception {
+        String remoteJarPath =
+                "hdfs://"
+                        + hdfsCluster.getURI().getHost()
+                        + ":"
+                        + hdfsCluster.getNameNodePort()
+                        + "/"
+                        + hdPath;
+
+        try (ClusterController clusterController = flink.startCluster(1)) {
+            Map<String, String> varsMap = new HashMap<>();
+            varsMap.put("$RESULT", this.result.toAbsolutePath().toString());
+            varsMap.put("$MODE", this.executionMode);
+            varsMap.put("$JAR_PATH", remoteJarPath);
+
+            List<String> sqlLines = initializeSqlLines(varsMap);
+
+            executeSqlStatements(clusterController, sqlLines);
+
+            checkJsonResultFile();
+        }
+    }
+
+    private List<String> initializeSqlLines(Map<String, String> vars) throws IOException {
```

Comment:
Here many duplicated method with `PlannerScalaFreeITCase`, so I think we can extract an base class to reuse the common method.